python/pyspark/sql/dataframe.py

```diff
@@ -109,15 +109,18 @@ def stat(self):
     @ignore_unicode_prefix
     @since(1.3)
     def toJSON(self, use_unicode=True):
-        """Converts a :class:`DataFrame` into a :class:`RDD` of string.
+        """Converts a :class:`DataFrame` into a :class:`DataFrame` of JSON string.
 
-        Each row is turned into a JSON document as one element in the returned RDD.
+        Each row is turned into a JSON document as one element in the returned DataFrame.
 
         >>> df.toJSON().first()
-        u'{"age":2,"name":"Alice"}'
+        Row(value=u'{"age":2,"name":"Alice"}')
         """
-        rdd = self._jdf.toJSON()
-        return RDD(rdd.toJavaRDD(), self._sc, UTF8Deserializer(use_unicode))
+        jdf = self._jdf.toJSON()
+        if self.sql_ctx._conf.pysparkDataFrameToJSONShouldReturnDataFrame():
+            return DataFrame(jdf, self.sql_ctx)
+        else:
+            return RDD(jdf.toJavaRDD(), self._sc, UTF8Deserializer(use_unicode))
```

Comment:
@ueshin, I think Scala side returns `Dataset[String]`, not DataFrame (`Dataset[Row]`). It is arguable because API usages will be different. For instance, 

```scala
scala> val df: DataFrame = Seq("a").toDF
df: org.apache.spark.sql.DataFrame = [value: string]

scala> df.foreach(println(_))
[a]

scala> val ds: Dataset[String] = Seq("a").toDS
ds: org.apache.spark.sql.Dataset[String] = [value: string]

scala> ds.foreach(println(_))
a
```

There's no concept of Dataset in PySpark side .. so if we should change for consistency reason, I doubt if we should change.

Reply:
Good point, but I'm feeling it's natural to return DataFrame because I think DataFrame in Python is a corresponding expression of DataFrame/Dataset in Scala/Java, whereas to return RDD is weird for me.
I can understand what you mean, and that is one of the reasons I added a config to restore the behavior, actually.
cc @gatorsmile @cloud-fan

Reply:
If scala side returns `Dataset[String]`, then I think it's reasonable for python side to return `DataFrame`, to be consistent with Scala side and supports all the relational operations(select, where, etc.) which `Dataset[String]` also supports.

Reply:
The problem is, the API usage is different. You can directly do, for instance, `map` on what `toJSON` returns for instance.

```python
df.map(lambda value: value + 1)
```

now it should be

```python
df.map(lambda row: row.value + 1)
```

This is pretty inconsistent with Scala side. I don't think we could say this is consistent since Dataset itself can be a replacement of an RDD but DataFrame is not.

For instance, `DataFrameReader.csv` or `DataFrameReader.json` doesn't accept `DataFrame` as an input because Scala side takes `Dataset[String]`. It only takes RDD.

```python
>>> spark.read.json(spark.range(1).rdd).show()
```
```
+---------------+
|_corrupt_record|
+---------------+
|      Row(id=0)|
+---------------+
```

```python
>>> spark.read.json(spark.range(1)).show()
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../spark/python/pyspark/sql/readwriter.py", line 293, in json
    raise TypeError("path can be only string, list or RDD")
TypeError: path can be only string, list or RDD
```

Reply:
BTW, if consistency is not the reason, I wouldn't fix. We can simply do:

```python
spark.createDataFrame(df.toJSON())
```

Reply:
I tested and updated the examples to self-contained.

Reply:
So which way is more consistent?

If we return DataFrame here, scala side can do `df.toJSON().map(value => value.xxx)` and python side needs to do `df.toJSON().map(lambda row: row.value.xxx)`.

If we return RDD here, scala side can do `df.toJSON().select(...)` and python needs to do `df.toJSON().toDF().select(...)`

I agree that there is no perfect way since pyspark has no dataset, we should pick a better but imperfect way.

Reply:
if both ways are not perfect, and no one is better than another, isn't it reasonable to add a config and let users themselves to pick?

Reply:
Also, in that way, we should fix `DataFrameReader.csv` or `DataFrameReader.json` to accept `DataFrame` because I believe the same argument can be there. It is also going to break one usecase, `spark.read.json(spark.range(1).toJSON())` without fixing `DataFrameReader.json`.

Reply:
I believe they can already pick one way by using other APIs. If this configuration is to control it, we shouldn't make this `true` by default BTW.

Reply:
That sounds interesting. Maybe we should fix `DataFrameReader.csv` and `DataFrameReader.json` to accept DataFrame of string in Python, regardless of the discussion here.
Let me try.

Reply:
I agree it's not clear which one is more consistent. I do not think we should add a config -- config options are, as someone once said, often a failure of design, that punts unresolved questions to the user. I think I'd slightly prefer keeping an RDD here if there's no compelling reason to significantly change the Pyspark API.

Reply:
Also, returning RDD and DataFrame are different. Returning RDD was the previous way even in Scala / Java side whereas DataFrame is not.

Keeping RDD as return has a legacy reason but returning DataFrame is basically we allow another incomplete way for `toJSON`.

Workaround is pretty easy anyway.

Reply:
We are trying to convince more end users to learn and use our high-level APIs [DataFrame/Dataset] instead of low-level APIs [RDD]. Returning RDDs encourages our end users to choose RDD APIs, like the example you pointed out
```
spark.range(1).toJSON().map(lambda value: value + "abc")
```

Using RDD APIs will be a blackbox and an barrier for our Spark SQL optimizer. Basically, we are unable to optimize user application in a more holistic way. 

For better perf, I think we should return a DataFrame instead of a RDD. cc @srowen @HyukjinKwon

Reply:
That's a reasonable position @gatorsmile if you're arguing to change the Scala `Dataset.toJSON` to return a `DataFrame` (really `Dataset[Row]`) also. Is that what you mean?

Reply:
For example, `limit` can be pushed down by our Spark SQL optimizer in theory, but I have not checked whether we can do it now. If not, we can improve it.

Reply:
If we're going to return DataFrame for Scala / Java side, yes. I'm okay.

If not, I think it can be arguable because it gives me an impression that we want to allow another incomplete way Python specifically due to optimization reason than usability - I also was thinking optimization shouldn't considerably affect user-side code in theory. Also, if not, we won't have consistency reason as well.

One difference is that we won't allow RDD-like APIs in PySpark side without another transformation to RDD since PySpark does not have dataset (RDD-like) specific APIs, for instance, `toJSON.map`. That's possible in Scala side but won't be able in PySpark side if we return DataFrame.

Reply:
BTW, there's workaround for `DataFrame` -> `DataFrame` to mimic `toJSON`. For instance,

```python
from pyspark.sql.functions import struct, to_json, col
df = spark.range(5)
df.select(struct(list(df)).alias("json")).select(to_json(col("json"))).show()
```